python/prima.py

```diff
@@ -0,0 +1,341 @@
+    None, ct.POINTER(ct.c_double), ct.POINTER(ct.c_double)
+)
+_prima_function_con_callback = ct.CFUNCTYPE(
+    None, ct.POINTER(ct.c_double), ct.POINTER(ct.c_double),
+    ct.POINTER(ct.c_double)
+)
+
+# prima_get_rc_string returns char*
+_handle.prima_get_rc_string.restype = ct.c_char_p
+
+# pure Python function converters
+
+
+def _prima_pure_callback(func, n):
+    def _inner(cx, cf):
+        x = [cx[i] for i in range(n)]
+        f = func(x)
+        cf[0] = f
+
+    return _prima_function_callback(_inner)
+
+
+def _prima_pure_con_callback(func, n, m):
+    def _inner(cx, cf, ccon):
+        x = [cx[i] for i in range(n)]
+        f, con = func(x)
+        cf[0] = f
+        for i in range(m):
+            ccon[i] = con[i]
+
+    return _prima_function_con_callback(_inner)
+
+
+@dataclass
+class Result:
+    """Optimization result."""
+    x: list
+    success: bool
+    status: int
+    message: str
+    fun: float
+    nfev: int
+
+
+def bobyqa(f, n, x0, xl, xu, rhobeg=1.0, rhoend=1e-3, maxfun=1000, iprint=0):
+    """
+    BOBYQA solver.
+
+    Parameters
+    ----------
+    f : callable
+        objective function, returns a float
+    n : int
+        input dimension
+    x0 : sequence of float
+        starting point
+    xl, xu : sequence of float
+        x bounds
+    rhobeg, rhoend : float
+        trust region radius
+    maxfun : int
+        max function evaluation
+    iprint : int
+        verbosity level
+
+    Returns
+    -------
+    res : Result
+        Result class
+    """
+    assert len(x0) == n
+    assert len(xl) == n
+    assert len(xu) == n
+    cn = ct.c_int(n)
+    cx = (ct.c_double * n)(*x0)
+    cf = ct.pointer(ct.c_double())
+    cxl = (ct.c_double * n)(*xl)
+    cxu = (ct.c_double * n)(*xu)
+    cnf = ct.pointer(ct.c_int())
+    crhobeg = ct.c_double(rhobeg)
+    crhoend = ct.c_double(rhoend)
+    cmaxfun = ct.c_int(maxfun)
+    ciprint = ct.c_int(iprint)
+    cfun = _prima_pure_callback(f, n)
+    status = _handle.prima_bobyqa(cfun, cn, cx, cf, cxl, cxu,
+            cnf, crhobeg, crhoend, cmaxfun, ciprint)
+    x = [cx[i] for i in range(n)]
+    success = (status == 0)
+    cstatus = ct.c_int(status)
+    message = _handle.prima_get_rc_string(cstatus).decode()
+    fun = cf[0]
+    nfev = cnf[0]
+    return Result(x, success, status, message, fun, nfev)
+
+
+def cobyla(m_nlcon, f, n, x0, m_ineq=0, Aineq=[], bineq=[], m_eq=0, Aeq=[], beq=[], xl=None, xu=None, rhobeg=1.0, rhoend=1e-3, maxfun=1000, iprint=0):
+    """
+    COBYLA solver.
+
+    Parameters
+    ----------
+    m_nlcon : int
+        non-linear constraint dimension
+    f : callable
+        objective function, returns a tuple consisting of a float
+        and the constraint values as a sequence of float of size m_nlcon
+    n : int
+        input dimension
+    x0 : sequence of float
+        starting point
+    m_ineq : int
+        Number of inequality constraints
+    m_eq : int
+        Number of equality constraints
```

Comment:
It seems that the arguments `Aineq`, `bineq`, `Aeq`, and `beq` are missing from the documentation.

Reply:
yes